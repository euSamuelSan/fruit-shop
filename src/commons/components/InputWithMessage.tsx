import React from 'react'
import styled from 'styled-components'
import Colors from '../Colors'
import bInput from './Input'

export type InputWithMessageProps = {
    placeholder?: string
    className?: string
    value: string
    onChange: (e: React.ChangeEvent<HTMLInputElement>) => void
    type?: string
    isInvalid: boolean
    setIsInvalid: (arg: boolean) => void
    invalidMessage: string
    label: string
}

export default function InputWithMessage({
    placeholder,
    className,
    value,
    onChange,
    type = 'text',
    isInvalid,
    setIsInvalid,
    invalidMessage,
    label,
}: InputWithMessageProps): JSX.Element {
    return (
        <Container className={className} onFocus={() => setIsInvalid(false)}>
            <Label>{label}</Label>
            <BasicInput
                type={type}
                placeholder={placeholder}
                value={value}
                onChange={onChange}
                isInvalid={isInvalid}
            />
            {isInvalid && <InvalidMessage>* {invalidMessage}</InvalidMessage>}
        </Container>
    )
}
const Container = styled.div`
    display: flex;
    flex-direction: column;
`

const BasicInput = styled(bInput)<Partial<InputWithMessageProps>>`
    outline: none;
    padding: 10px;
    border: 2px solid;
    border-radius: 5px;
    border-color: ${props =>
        props.isInvalid ? Colors.strawberry : Colors.lightBlueGrey};
    color: ${Colors.black};
`

const InvalidMessage = styled.p`
    font-size: 12px;
    color: ${Colors.strawberry};
    margin: 3px 0 0 5px;
`

const Label = styled.label`
    color: ${Colors.black};
    font-weight: 600;
    padding: 5px;
`
