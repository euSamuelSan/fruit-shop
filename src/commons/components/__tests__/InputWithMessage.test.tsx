import { fireEvent, render, RenderResult, screen } from '@testing-library/react'

import InputWithMessage, { InputWithMessageProps } from '../InputWithMessage'

describe('InputWithMessage', () => {
    const getProps = (): InputWithMessageProps => ({
        value: '',
        onChange: jest.fn(),
        isInvalid: false,
        setIsInvalid: jest.fn(),
        invalidMessage: 'ANY_INVALID_MESSAGE',
        label: 'ANY_LABEL',
    })

    const renderInputWithMessage = (props = getProps()): RenderResult => {
        return render(<InputWithMessage {...props} />)
    }

    it('Deve desenhar o componente corretamente', () => {
        renderInputWithMessage()

        expect(document.body).toMatchSnapshot()
    })

    it('Deve desenhar o componente com o Input invÃ¡lido', () => {
        renderInputWithMessage({
            ...getProps(),
            isInvalid: true,
        })

        expect(document.body).toMatchSnapshot()
    })

    it('Deve chamar o onChange ao alterar o valor do input', () => {
        const props = {
            ...getProps(),
            placeholder: 'ANY_PLACEHOLDER',
        }

        renderInputWithMessage(props)
        const input = screen.getByPlaceholderText('ANY_PLACEHOLDER')

        fireEvent.change(input, {
            target: {
                value: 'ANY_PLACEHOLDER',
            },
        })

        expect(props.onChange).toHaveBeenCalled()
    })
})
