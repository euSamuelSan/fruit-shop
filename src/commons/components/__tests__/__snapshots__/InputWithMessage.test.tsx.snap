// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputWithMessage Deve desenhar o componente com o Input inv√°lido 1`] = `
<body>
  <div>
    <div
      class="sc-gsnTZi dxCewW"
    >
      <label
        class="sc-eCYdqJ kjGVhh"
      >
        ANY_LABEL
      </label>
      <input
        class="sc-bczRLJ hKuxkL sc-dkzDqf enGFyK"
        type="text"
        value=""
      />
      <p
        class="sc-hKMtZM kNePtT"
      >
        * 
        ANY_INVALID_MESSAGE
      </p>
    </div>
  </div>
</body>
`;

exports[`InputWithMessage Deve desenhar o componente corretamente 1`] = `
<body>
  <div>
    <div
      class="sc-gsnTZi dxCewW"
    >
      <label
        class="sc-eCYdqJ kjGVhh"
      >
        ANY_LABEL
      </label>
      <input
        class="sc-bczRLJ hKuxkL sc-dkzDqf dsjBsf"
        type="text"
        value=""
      />
    </div>
  </div>
</body>
`;
